pkg/agent/svid/rotator.go

```diff
@@ -0,0 +1,162 @@
+package svid
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/imkira/go-observer"
+	"github.com/spiffe/spire/pkg/agent/client"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/api/node"
+)
+
+type Rotator interface {
+	Start()
+	Stop()
+
+	State() State
+	Subscribe() observer.Stream
+}
+
+type rotator struct {
+	c      *RotatorConfig
+	stop   chan struct{}
+	done   chan struct{}
+	client client.Client
+
+	state observer.Property
+
+	m       sync.RWMutex
+	running bool
+	// Mutex used to protect access to c.BundleStream.
+	bsm *sync.RWMutex
+}
+
+type State struct {
+	SVID *x509.Certificate
+	Key  *ecdsa.PrivateKey
+}
+
+// Start starts the rotator.
+func (r *rotator) Start() {
+	go r.run()
+}
+
+func (r *rotator) Stop() {
+	close(r.stop)
+	if r.isRunning() {
+		<-r.done
+		r.setRunning(false)
+	}
+}
+
+func (r *rotator) State() State {
+	return r.state.Value().(State)
+}
+
+func (r *rotator) Subscribe() observer.Stream {
+	return r.state.Observe()
+}
+
+// run
+// - Starts a ticker which monitors the server SVID for expiration and invokes
+// rotateSVID() as necessary.
+// - Reads the next trust bundle received on BundleStream.
+func (r *rotator) run() {
+	r.setRunning(true)
+	t := time.NewTicker(r.c.Interval)
+
```

Comment:
defer t.Stop() to prevent leaking a goroutine

Reply:
:+1: